dvc/remote/s3.py

```diff
@@ -209,8 +209,15 @@ def list_cache_paths(self):
 
     def exists(self, path_info):
         dir_path = path_info / ""
-        fname = next(self._list_paths(path_info, max_items=1), "")
-        return path_info.path == fname or fname.startswith(dir_path.path)
+
+        if self.isdir(dir_path):
+            return True
+
+        for fname in self._list_paths(path_info):
+            if path_info.path == fname:
+                return True
```

Comment:
I am concerned about performance if the user adds root folder, so, recursively checking for `exists()` could be a huge slowdown. I'm quite not sure if it is cached. 

cc @efiop

Reply:
You still need to check for starts with, a dir file might be absent, but did still counts as existing. And why do you check for issue?

Reply:
> a dir file might be absent
I'm wondering when those cases occur, as, both `isdir` and `exists` is using `_list_paths()`.

> And why do you check for issue?
I don't understand. Please clarify.

Reply:
Why do you check for `isdir()`? Sorry that was from phone)

Reply:
> I'm wondering when those cases occur

That might occur if someone is adding external dependency, which user created with unknown method. Or someone might simply remove the dir file.

Reply:
Say,  I have a file `data/empty_file`. If I check for `exists('data'), what should happen? In GS and S3, it's all prefix, so, before checking for exact match, it's good to check if it's a directory path.

But, talking to you, in GS and S3, we can have a file with paths `<file>/` and `<file>` as different files. But, this is an edge-case, that, I'd say, we ignore for now (as does the `gsutil`).

Reply:
>  Or someone might simply remove the dir file.

`isdir` does not check for an exact match, so, if the prefix matches, then, it won't be a problem.

Reply:
If we have `data/empty_file` then `.isdir('data')` and `.exists('data')` should both return true, this is the reason for `.starswith()`.

Reply:
This is covered by `isdir()` though. I wonder whether we can avoid doing 2 list calls to check for existence, we make lots of exists calls and most for the cache files for which we don't care about dirs.

Reply:
> what should happen

If we have data/empty_file then .isdir('data') and .exists('data') should both return true, this was the reason for .starswith().

Reply:
BTW, the original idea about using listing for exists was caused by it being faster than head request which was an effect of [this bug](https://github.com/boto/boto3/issues/1128) of urlllib3 vendored within boto. However, the bug was [fixed in urllib3](https://github.com/urllib3/urllib3/pull/1235), which was later [unvendored from boto](https://github.com/boto/botocore/pull/1300). So we should reconsider using `.head_object()`.

BTW, `.head_object()` was faster when key exists [2.5 years ago](https://www.peterbe.com/plog/fastest-way-to-find-out-if-a-file-exists-in-s3), so I wonder whether this was a good idea at any point). Even when if we find that `list_objects` is faster for us than `head_objects` then we might want to still use the latter to never run into big lists.

Additionally we use listing for exists in GS, Azure and OSS, I don't know why. Were this simply an apishness, i.e. a copy paste? :) We should really start commenting our implementation, especially when using hacks like this.

Additional consideration about cache, we don't really want to check `.isdir()` for every checksum, but we want `.exists()` to work for both files and dirs for remotes supporting that. This contradiction might be resolved by calling `.isfile()` instead of `.exists()` inside `.cache_exists()`. We will need to make `isfile()` work properly though, now it is just a stab in `RemoteBASE`.

Reply:
@Suor To check if dir exists, we actually need to list, as we can't know what specific file to target in the head_object. E.g. `dir/.` objest that we use for creating empty dirs is not guaranteed to exist. So we need to do a list on `dir/` to quickly know if there is something inside or not. Note that we are passing max items = 1 in the current implementation, as we don't need to list all paths.

Reply:
@Suor You have a great point about head_object, indeed, looks like we can start using it again these days. So we could implement `isfile()` with it and then just implement `exists` as 
```
if self.isfile(path_info):
    return True

return self.isdir(path_info)
```
which is neat 🙂

Reply:
So, should I implement `isfile()` and use `get_head_object()` and do as @efiop mentioned for `.exists()`?

Reply:
I did it in da0d9e8. I haven't changed `.cache_exists()` to use `isfile()`, as I don't want to conflict with the other PR. Probably, in the next PR.

Reply:
@skshetry you can't simply change `.exists()` to `.isfile()` in `.cached_exists()` because of two reasons:
1. It's not implemented properly for all remotes.
2. It might be slower for some remotes.

So there is an area for discussion still. This should definitely be a separate PR. You may create an issue though :)

Reply:
@Suor, okay, I'll create an issue, but can you provide me more information about how caching works here? Sorry, if this sounds naive, there is a directory checksum, so, is that not used, or has different use?

Reply:
This is about dvc cache. When any data files get under dvc control we copy them to some checksum named files, like `.dvc/cache/ab/1hj3....`. Then you `dvc push` those cache files are uploaded to some remote, say S3 or GS. Then later on some ops like `dvc pull/push/status -r` we need to compare what we have locally and on the remote. We do that with hordes of exists calls, which don't care about that being dirs. We have directory checksum files, they are files not dirs too, this a separate story though.

Reply:
Created #2877 for `cache_exists()`